File: RayTracer/ViewPlane.cs
using System;

namespace RayTracer
{
    public class ViewPlane
    {
        public static ViewPlane Instance;
        double worldWidth;
        double worldHeight;
        int pixelHeight;
        int pixelWidth;
        Point3 position;
        Point3 upperLeft;
        Point3 unitRight;
        Point3 unitDown;

        public int PixelWidth
        {
            get { return pixelWidth; }
            set { pixelWidth = value; }
        }

        public int PixelHeight
        {
            get { return pixelHeight; }
            set { pixelHeight = value; }
        }


        public ViewPlane(int width, int height)
        {
            pixelWidth = width;
            pixelHeight = height;


            worldHeight = 2.0 * Math.Tan((Camera.Instance.FieldOfView / 2f) * Math.PI / 180.0);
            worldWidth = worldHeight * (double)((double)pixelWidth / (double)PixelHeight);

            PreCalculate();

        }


        void PreCalculate()
        {
            upperLeft = GetUpperLeft();
            unitRight = (Camera.Instance.U.Value * (worldWidth / (Double)PixelWidth) * -1);
            unitDown = (Camera.Instance.V.Value * (worldHeight / (Double)PixelHeight) * -1);
        }

        Point3 GetUpperLeft()
        {

            Point3 center = Camera.Instance.Position;
            center += Camera.Instance.W.Value;
            position = center;
            Point3 upperLeft =
                center
                + Camera.Instance.U.Value * (worldWidth / 2.0)    //U is left
                + Camera.Instance.V.Value * (worldHeight / 2.0);  //V is up


            return upperLeft;
        }

        public Point3 GetNewLocation(int col, int row)
        {
            Point3 newLocation =
                upperLeft
                + unitRight * (col + .5)
                + unitDown * (row + .5);
            return newLocation;
        }

        public Point3 GetNewLocation(int col, int row, Double col2, Double row2)
        {

            Point3 newLocation =
                upperLeft
                + unitRight * (col + .5 + col2)
                + unitDown * (row + .5 + row2);
            return newLocation;
        }


        public void ShowInformation()
        {
            Console.WriteLine("View Plane Information =============================");
            Console.WriteLine("Upper Left");
            upperLeft.ShowInformation();
            Console.WriteLine("Center");
            position.ShowInformation();
            Console.WriteLine("H / W pixel : " + PixelHeight + " / " + PixelWidth);
            Console.WriteLine("H / W world : " + worldHeight + " / " + worldWidth);
            Console.WriteLine("====================================================");
        }

    }
}

File: RayTracer - CUDA/RayTracerCUDA/ViewPlane.cpp
#include "ViewPlane.h"



ViewPlane & ViewPlane::Instance()
{
	static ViewPlane inst;
	return inst;
}

void ViewPlane::Init(int width, int height)
{
	pixelH = height;
	pixelW = width;
	worldH = 2.f * tanf((Camera::Instance().fov / 2.f) * (float)M_PI / 180.f);
	worldW = worldH * (pixelW / (float)pixelH);

	Precalculate();

}

void ViewPlane::Precalculate()
{
	upperLeft = getUpperLeft();
	unitRight = ((Camera::Instance().U * (worldW / (float)pixelW)) * -1.f);
	unitRight = ((Camera::Instance().V * (worldH / (float)pixelH)) * -1.f);
}

Point3 & ViewPlane::getUpperLeft()
{
	Point3 c = Camera::Instance().pos;
	c = c + Camera::Instance().W;
	pos = c;
	Point3 upperLeft =
		c
		+ Camera::Instance().U * (worldW * .5f)
		+ Camera::Instance().V * (worldH * .5f);

		return upperLeft;
}

Point3& ViewPlane::getNewLocation(int col, int row)
{
	Point3 uleft = upperLeft;
	Vec3 uRight = unitRight;
	Vec3 uDown = unitDown;

	return		uleft
		+ uRight * (col + .5f)
		+ uDown * (row + .5f)
		;

}

ViewPlane::ViewPlane()
{
}

ViewPlane::ViewPlane(const ViewPlane &)
{
}

ViewPlane & ViewPlane::operator=(const ViewPlane &)
{
	return Instance();
}


ViewPlane::~ViewPlane()
{
}
